### PR TITLE
Move Task to shared partition

### PR DESCRIPTION
**Pull request by marek-safar**
Contributes to #21932

### CODE REVIEW

File: src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+namespace System.Threading.Tasks
+{
+    public partial class Task : IAsyncResult, IDisposable
+    {
+        static partial void EtwNewID(int newId)
+        {
+            TplEtwProvider.Log.NewID(newId);
+        }
+
+        partial void EtwTaskStarted(Task previousTask, ref Guid savedActivityID)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: I'm a little concerned that this factoring is going to impact performance, as most of these methods are not going to get inlined automatically, and we're now reading the static log multiple times where previously it was read only once.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 2]
```diff
@@ -882,21 +900,10 @@ internal bool MarkStarted()
             return AtomicStateUpdate(TASK_STATE_STARTED, TASK_STATE_CANCELED | TASK_STATE_STARTED);
         }
 
-        internal bool FireTaskScheduledIfNeeded(TaskScheduler ts)
+        [Conditional("CORECLR")]
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Why is this needed?

Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: It's not, I'm just not sure if CoreRT runtime can remove the call

File: src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs (hunk 1 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: I'd expect the read from the static read-only field to be inlined. If you are concern about extra call maybe aggressive inlining would help but not sure if the call price is even measurable in this context.

Comment (filipnavara) [reply src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: It usually is inlined and eliminated, but it depends on the compilation tier. Explicit aggressive inlining or aggressive optimization may actually hurt the performance. I think it has to do with the fact that a static read-only field may not be assigned by the time the first JIT compilation is performed, but if you let tiering do its thing then later recompilation will see it assigned and optimize properly. Whatever the reason I suggest not to do it without a benchmark backing it up.

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: This split should not be needed. It should be fine to have same TplEtwProvider  as is in CoreCLR. TplEtwProvider is in CoreRT already; it just factored differently for historic reasons.

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/Internal/Threading/Tasks/Tracing/TaskTrace.cs and https://github.com/dotnet/corert/blob/master/src/System.Private.Threading/src/Internal/Threading/TaskTraceCallbacksImplementation.cs can be deleted in CoreRT; and https://github.com/dotnet/corert/blob/master/src/System.Private.Threading/src/System/Threading/Tasks/TPLETWProvider.cs can be moved to CoreLib and unified with the CoreCLR one.

Comment (marek-safar) [reply src/System.Private.CoreLib/src/System/Threading/Tasks/Task.CoreCLR.cs]: @jkotas done (will make TplEtwProvider changes in different PR). I kept the Etw separation for easier reading and I don't think it should cause any perf issue.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 3]
```diff
@@ -3253,19 +3227,14 @@ internal void FinishContinuations()
         {
             Debug.Assert(continuationObject != null);
 
-            TplEtwProvider etw = TplEtwProvider.Log;
-            bool tplEtwProviderLoggingEnabled = etw.IsEnabled();
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: I know @benaadams and others have been fine-tuning the performance of this code. Caching of `Log.IsEnabled` in a local looks like the kind of change that would be done as part of the fine-tuning. Are we going to lose any cycles when the caching is removed?

Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Caching it in local variable make it much harder for illinker to remove the code

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: `.Log` is a static readonly based `EventSource` of the sealed type `TplEtwProvider`; and `IsEnabled()` is a non-virtual => `m_eventSourceEnabled` so it should inline happily. 

Main impact here seems to be 1 additional read vs having the local (as they are pairs); so it probably wouldn't be much.


Not used here, but `DiagnosticSource` is a more problematic type as its `.IsEnabled` is virtual so caching its result is helpful; although using its derived type `DiagnosticListener` instead is better as it has a fast non-virtual `.IsEnabled()`.

Removing `ThreadPool`'s caching of the Log states in the `Dispatch` loop might be more of an issue though (or any `.IsEnabled(string,..)`) vairant where its doing more checking; which here are guarded by the fast `.IsEnabled()` checks.

Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Long story short, I'll defer to @stephentoub ;)

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 4]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
             }
         }
 
+        #region EventSource Tracing
+
+        void EtwTaskStarted(Task previousTask, ref Guid savedActivityID)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: All of these void Etw methods should be explicitly private.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 5]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
...
+            if (etwLog.TasksSetActivityIds)
+                EventSource.SetCurrentThreadActivityId(TplEtwProvider.CreateGuidForTaskID(this.Id), out savedActivityID);
+            // previousTask holds the actual "current task" we want to report in the event
+            if (previousTask != null)
+                etwLog.TaskStarted(previousTask.m_taskScheduler.Id, previousTask.Id, this.Id);
+            else
+                etwLog.TaskStarted(TaskScheduler.Current.Id, 0, this.Id);
+        }
+
+        void EtwTaskCompleted(Task previousTask, ref Guid savedActivityID)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Why did all of these need to be refactored into helpers?

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 6]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
...
+            {
+                TplEtwProvider etwLog = TplEtwProvider.Log;
+                etwLog.AwaitTaskContinuationScheduled(TaskScheduler.Current.Id, Task.CurrentId ?? 0, continuationTask.Id);
+            }
+        }
+
+        void EtwRunningContinuationList(int index, object obj)
+        {
+            TplEtwProvider etw = TplEtwProvider.Log;
+            etw.RunningContinuationList(Id, index, obj);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: This can just be:
```C#
TplEtwProvider.Log.RunningContinuationList(Id, index, obj);
```

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 7]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
             }
         }
 
+        #region EventSource Tracing
+
+        void EtwTaskStarted(Task previousTask, ref Guid savedActivityID)
+        {
+            TplEtwProvider etwLog = TplEtwProvider.Log;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: As long as we're touching all of this event-related code and moving stuff around, can we rename TplEtwProvider to instead be TplEventSource?

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 8]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
...
+            etw.RunningContinuation(Id, continuationObject);
+        }
+
+        void EtwAwaitTaskContinuationScheduled(Task continuationTask)
+        {
+            if ((this.Options & (TaskCreationOptions)InternalTaskOptions.PromiseTask) != 0 && !(this is ITaskCompletionAction))
+            {
+                TplEtwProvider etwLog = TplEtwProvider.Log;
+                etwLog.AwaitTaskContinuationScheduled(TaskScheduler.Current.Id, Task.CurrentId ?? 0, continuationTask.Id);
+            }
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Same as below... the local etwLog isn't needed

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 9]
```diff
@@ -700,6 +704,97 @@ private string DebuggerDisplayMethodDescription
...
+                Task parentTask = m_contingentProperties?.m_parent;
+                etwLog.TaskScheduled(ts.Id, currentTask == null ? 0 : currentTask.Id,
+                                     this.Id, parentTask == null ? 0 : parentTask.Id, (int)this.Options);
+            }
+        }
+
+        void EtwRunningContinuation(object continuationObject)
+        {
+            TplEtwProvider etw = TplEtwProvider.Log;
+            etw.RunningContinuation(Id, continuationObject);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Same as below... the local etw isn't needed... and when we do have a local, it'd be nice to name it consistently, e.g. etw vs etwLog

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 5 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: It's not needed for this PR it's just way how the PR evolved and I found it useful to have it separated similar way as other tracing mechanisms in Task. Some of them could possibly be reused by TaskContinuation but I didn't look into that yet.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 10]
```diff
@@ -2286,45 +2362,6 @@ internal void AddExceptionsFromChildren(ContingentProperties props)
...
-        {
-            // If the task has marked itself as Completed, then it either a) already observed this exception (so we shouldn't handle it here)
-            // or b) completed before the exception ocurred (in which case it shouldn't count against this Task).
-            if (!IsCompleted)
-            {
-                HandleException(tae);
-                FinishThreadAbortedTask(delegateRan: false);
-            }
-        }
-
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: We only kept it around for Mono.  If Mono no longer needs it, cleaning it up sounds good to me.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 11]
```diff
@@ -2618,7 +2626,7 @@ public ConfiguredTaskAwaitable ConfigureAwait(bool continueOnCapturedContext)
             if (tc != null)
             {
                 if (!AddTaskContinuation(tc, addBeforeOthers: false))
-                    tc.Run(this, bCanInlineContinuationTask: false);
+                    tc.Run(this, canInlineContinuationTask: false);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: This naming has bugged me for years.  Thanks for fixing it :)

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: The caching in a local was done years ago, before IsEnabled could be devirtualized, and it saved a static read in various places.  If you can confirm that it IsEnabled previously wasn't being devirtualized and now is being devirtualized and inlined, I'm happy for the change, as that'll have a net benefit. If it was already getting devirtualized/inlined, then we should run some perf tests just to confirm that there's no measurable negative impact, though I suspect there won't be.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 12]
```diff
@@ -3378,7 +3345,7 @@ private void RunOrQueueCompletionAction(ITaskCompletionAction completionAction,
             }
         }
 
-        private void LogFinishCompletionNotification()
+        private static void LogFinishCompletionNotification()
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Since you've created a bunch of new logging-related methods, it'd be great to group them all together in the same place in the file.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 13]
```diff
@@ -6110,6 +6070,9 @@ internal static Task<TResult> CreateUnwrapPromise<TResult>(Task outerTask, bool
             return new UnwrapPromise<TResult>(outerTask, lookForOce);
         }
 
+#if CORERT
+        [DependencyReductionRoot]
+#endif
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: @jkotas, rather than having ifdefs around DependencyReductionRoot, could we define one in coreclr?  It could even be attributed as Conditional so that it's compiled away in release builds.

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs [hunk 14]
```diff
@@ -652,17 +652,6 @@ internal static bool Dispatch()
...
-                // This is here to catch the case where this thread is aborted between the time we exit the finally block in the dispatch
-                // loop, and the time we execute the work item.  QueueUserWorkItemCallback uses this to update its accounting of whether
-                // it was executed or not (in debug builds only).  Task uses this to communicate the ThreadAbortException to anyone
-                // who waits for the task to complete.
-                //
-                if (outerWorkItem is Task task)
-                {
-                    task.MarkAbortedFromThreadPool(tae);
-                }
-
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: Do we need this catch block at all?
cc: @kouvel

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: > `IsEnabled` could be devirtualized

IsEnable is not virtual. I do not think devirtualization has anything to do with this.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: You're right; I was thinking of DiagnosticSource.IsEnabled rather than EventSource.IsEnabled.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: In which case, I would like to see some validation done that this doesn't regress anything.  I don't expect it will measurable, but still...

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 13 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: There are not that many and most of them are in places that are different between CoreCLR and ProjectN for other reasons.

If somebody want to do it, I would not mind, but I am not sure whether it is worth it. I would love for the `DependencyReductionRoot` attribute to go away eventually.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 15]
```diff
@@ -6110,6 +6070,9 @@ internal static Task<TResult> CreateUnwrapPromise<TResult>(Task outerTask, bool
             return new UnwrapPromise<TResult>(outerTask, lookForOce);
         }
 
+#if CORERT
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: This should be `#if PROJECTN`. These attributes are just for .NET Native. CoreRT does not need them.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 16]
```diff
@@ -6171,6 +6134,9 @@ internal static Delegate[] GetDelegatesFromContinuationObject(object continuatio
             return null;
         }
 
+#if CORERT
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Same here

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 7 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Right.

Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: @stephentoub I am fine with rename but would prefer to do it once all code using it is in shared

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Ok

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 3 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: I'll undo the cleanup, the whole etw setup in Task looks very fishy to me

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 15 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Done

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs [hunk 17]
```diff
@@ -652,17 +652,6 @@ internal static bool Dispatch()
             }
             catch (ThreadAbortException tae)
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: This catch block should be removed altogether - it can be done in follow up change.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 18]
```diff
@@ -882,21 +886,17 @@ internal bool MarkStarted()
             return AtomicStateUpdate(TASK_STATE_STARTED, TASK_STATE_CANCELED | TASK_STATE_STARTED);
         }
 
-        internal bool FireTaskScheduledIfNeeded(TaskScheduler ts)
+        internal void EtwTaskScheduled(TaskScheduler ts)
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: The original name for this (`FireTaskScheduledIfNeeded`) looked better.

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs (hunk 17 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: Will leave that to @filipnavara who is working on TP

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 18 above)
Comment (marek-safar) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: ok

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs (hunk 17 above)
Comment (filipnavara) [reply src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs]: Already done in my last iteration of the ThreadPool PR.

**Comment by jkotas** [comment]
@Anipik Could you please take a look at the CoreLib mirror? It does not seem to be picking up this change.

**Comment by Anipik** [comment]
Done :)
